src/orb.jl

```diff
@@ -0,0 +1,118 @@
+        next_w = ceil(Int, prev_w / downsample)
+        img_smoothed = imfilter_gaussian(prev, [sigma, sigma])
+        img_scaled = Images.imresize(img_smoothed, (next_h, next_w))
+        push!(pyramid, img_scaled)
+        prev = img_scaled
+        prev_h, prev_w = size(img_scaled)
+    end
+    pyramid
+end 
+
+function create_descriptor{T<:Gray}(img::AbstractArray{T, 2}, params::ORB)
+    pyramid = gaussian_pyramid(img, params.levels, params.downsample, params.sigma)
+    keypoints_stack = map(image -> Keypoints(fastcorners(image, params.n_fast, params.threshold)), pyramid)
+    patch = ones(31, 31)
+    orientations_stack = map((image, keypoints) -> corner_orientations(image, keypoints, patch), pyramid, keypoints_stack)
+    harris_response_stack = map(image -> harris(image, k = params.harris_factor), pyramid)
+    descriptors = BitArray[]
+    ret_keypoints = Keypoint[]
+    harris_responses = Float64[]
+    scales = Float64[]
+
+    for i in 1:params.levels + 1
+        desc, ret_key = create_descriptor(pyramid[i], keypoints_stack[i], orientations_stack[i], params)
+        ret_key_scaled = [Keypoint(floor(Int, rk[1] * (params.downsample ^ (i - 1))), floor(Int, rk[2] * (params.downsample ^ (i - 1)))) for rk in ret_key]
+        append!(descriptors, desc)
+        append!(ret_keypoints, ret_key_scaled)
+        append!(harris_responses, harris_response_stack[i][ret_key])
+        append!(scales, ones(length(ret_key)) * floor(Int, (params.downsample ^ (i - 1))))
+    end
+    
+    if params.num_keypoints < length(descriptors)
+        first_n_indices = selectperm(harris_responses, 1:params.num_keypoints, rev = true)
+        return descriptors[first_n_indices], ret_keypoints[first_n_indices], scales[first_n_indices]
+    end
+
+    descriptors, ret_keypoints, scales
+end
+
+function create_descriptor{T<:Gray}(img::AbstractArray{T, 2}, keypoints::Keypoints, orientations::Array{Float64}, params::ORB)
+    descriptors = BitArray[]
+    ret_keypoints = Keypoint[]
+    for (i, k) in enumerate(keypoints)
+        orientation = orientations[i]
+        sin_angle = sin(orientation)
+        cos_angle = cos(orientation)
+        descriptor = BitArray([])
+        for (y0, x0, y1, x1) in sampling_pattern
+            pixel0 = CartesianIndex(floor(Int, sin_angle * y0 + cos_angle * x0), floor(Int, cos_angle * y0 - sin_angle * x0))
+            pixel1 = CartesianIndex(floor(Int, sin_angle * y1 + cos_angle * x1), floor(Int, cos_angle * y1 - sin_angle * x1))
+            checkbounds(Bool, img, k + pixel0) && checkbounds(Bool, img, k + pixel1) || @goto endofloop
+            push!(descriptor, img[k + pixel0] < img[k + pixel1])
+        end
+        push!(ret_keypoints, k)
+        push!(descriptors, descriptor)
+        @label endofloop
+    end
+    descriptors, ret_keypoints
+end
+
+const sampling_pattern = [[8, -3, 9, 5], [4, 2, 7, -12], [-11, 9, -8, 2], [7, -12, 12, -13], [2, -13, 2, 12], [1, -7, 1, 6], 
```

Comment:
Will reading them from a text file each time a descriptor needs to be calculated take up too much time?